File: 243. Shortest Word Distance.cpp
class Solution {
public:
    // Two pointers: O(N)time, O(1)space, 81.01%, ok
    int shortestDistance(vector<string>& words, string word1, string word2) {
        if(words.empty()) return -1;
        int p1 = -1, p2 = -1;
        int ret = INT_MAX;
        for(int i = 0; i < words.size(); i++){
            if(words[i] == word1){
                if(p2 != -1) ret = min(ret, abs(i - p2));
                p1 = i;
            }
            if(words[i] == word2){
                if(p1 != -1) ret = min(ret, abs(i - p1));
                p2 = i;
            }
        }
        return ret;
    }

    // more elegant, Two pointers: O(N)time, O(1)space, 81.01%, ok
    int shortestDistance(vector<string>& words, string word1, string word2) {
        if(words.empty()) return -1;
        int p1 = -1, p2 = -1;
        int ret = INT_MAX;
        for(int i = 0; i < words.size(); i++){
            if(words[i] == word1) p1 = i;
            if(words[i] == word2) p2 = i;
            if(p1 != -1 && p2 != -1)
                ret = min(ret, abs(p1 - p2));
        }
        return ret;
    }
};

File: 243. Shortest Word Distance.py
class Solution(object):
    
    # brute force: O(N2) time, O(1) space
    def shortestDistance(self, words, word1, word2):
        """
        :type words: List[str]
        :type word1: str
        :type word2: str
        :rtype: int
        """
        ret = float('inf')
        for i in xrange(len(words)):
            if words[i] == word1:
                for j in xrange(len(words)):
                    if words[j] == word2:
                        ret = min(ret, abs(j - i))
        return ret
    
    # one pass: O(N) time, O(1) space
    def shortestDistance(self, words, word1, word2):
        index1, index2, ret = -1, -1, float('inf')
        for i, word in enumerate(words):
            if word == word1:
                index1 = i
            if word == word2:
                index2 = i
            if index1 != -1 and index2 != -1:
                ret = min(abs(index1 - index2), ret)
        return ret
        